File: Week_02/G20200343030632/Leetcode_22_632.java
import java.util.LinkedList;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new LinkedList<>();
        helper(n, 0, 0, "", res);
        return res;
    }
    private void helper(int n, int left, int right, String s, List<String> res) {
        if (left == n && right == n) {
            res.add(s);
            return;
        }
        if (left < n) {
            helper(n, left + 1, right, s + "(", res);
        }
        if (right < left) {
            helper(n, left, right + 1, s + ")", res);
        }
    }
}

File: Week_02/G20200343030632/Leetcode_22_632.py



class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        self.helper(n, 0, 0, '', res)
        return res
    
    def helper(self, n: int, left: int, right: int, s, res):
        if left == n and right == n:
            res.append(s)
            return 
        
        if left < n:
            self.helper(n, left + 1, right, s + '(', res)
        if right < left:
            self.helper(n, left, right + 1, s + ')', res)
        
